lib/kitchen/driver/gce.rb

```diff
@@ -16,154 +17,503 @@
         end
-        gen_name
       end
 
-      def select_zone
-        if config[:region] == 'any'
-          zone_regexp = /^[a-z]+\-/
-        else
-          zone_regexp = /^#{config[:region]}\-/
+      def migrate_setting
+        config[:auto_migrate] ? "MIGRATE" : "TERMINATE"
+      end
+
+      def instance_service_accounts
+        return if config[:service_account_scopes].nil? || config[:service_account_scopes].empty?
+
+        service_account        = Google::Apis::ComputeV1::ServiceAccount.new
+        service_account.email  = config[:service_account_name]
+        service_account.scopes = config[:service_account_scopes].map { |scope| service_account_scope_url(scope) }
+
+        Array(service_account)
+      end
+
+      def service_account_scope_url(scope)
+        return scope if scope.start_with?("https://www.googleapis.com/auth/")
+        "https://www.googleapis.com/auth/#{translate_scope_alias(scope)}"
+      end
+
+      def translate_scope_alias(scope_alias)
+        SCOPE_ALIAS_MAP.fetch(scope_alias, scope_alias)
+      end
+
+      def instance_tags
+        Google::Apis::ComputeV1::Tags.new.tap { |tag_obj| tag_obj.items = config[:tags] }
+      end
+
+      def wait_time
+        config[:wait_time]
+      end
+
+      def refresh_rate
+        config[:refresh_rate]
+      end
+
+      def wait_for_status(requested_status, &block)
+        last_status = ""
+
+        begin
+          Timeout.timeout(wait_time) do
+            loop do
+              item = block.call
+              current_status = item.status
+
+              unless last_status == current_status
+                last_status = current_status
+                info("Current status: #{current_status}")
+              end
+
+              break if current_status == requested_status
+
+              sleep refresh_rate
+            end
+          end
+        rescue Timeout::Error
+          error("Request did not complete in #{wait_time} seconds. Check the Google Cloud Console for more info.")
+          raise
         end
-        zones = connection.zones.select do |z|
-          z.status == 'UP' && z.name.match(zone_regexp)
+      end
+
+      def wait_for_operation(operation)
+        operation_name = operation.name
+
+        wait_for_status("DONE") { zone_operation(operation_name) }
+
+        errors = operation_errors(operation_name)
+        return if errors.empty?
+
+        errors.each do |error|
+          error("#{error.code}: #{error.message}")
         end
-        fail 'No up zones in region' unless zones.length >= 1
-        zones.sample.name
+
+        raise "Operation #{operation_name} failed."
       end
 
-      def wait_for_up_instance(instance, state)
-        instance.wait_for do
-          print '.'
-          ready?
+      def wait_for_server
+        begin
+          instance.transport.connection(state).wait_until_ready
+        rescue
+          error("Server not reachable. Destroying server...")
```

Comment:
I'm on the fence about this, but if the user set `autodelete_disk` to false, do you want to warn the user that they've "leaked" a disk resource?